### DISCUSSION TITLE
Plotting surface with cyclic colormap

### DISCUSSION THREAD

**Discussion by davidcortesortuno** [pyvista/pyvista#5342] [Q&A] [answered]
Hi, what's the best way to plot a surface, using the `cmap=` and `scalars=` arguments, using a cyclic colormap? If I use Matplotlib's `hsv` for example, and my data is normalized from `0` to `2*pi`, neighbouring points with values at the extrema of the data range, i.e. `0` and `2*pi`, are interpolated wrongly, so I observe awful artifacts. Is there a way to control the interpolation used to plot the colours?

This issue is similar to the one here https://community.plotly.com/t/interpolating-a-cyclic-colormap/53365

**Comment by darikg** [comment]
Is it enough to set `interpolate_before_map=False`? That will at least interpolate between colors, rather than between scalar values

**Reply by davidcortesortuno** [reply]
This significantly improves the result :) However, I still see some minor artifacts, I guess there is still some interpolation around

![image](https://github.com/pyvista/pyvista/assets/12200541/0c1a173a-0dd6-4680-bae2-ef77eda45249)

**Reply by darikg** [reply]
Just a thought -- maybe it's possible to do some manual interpolation like this, assuming the original scalar property is called 'angle'

```python
import numpy as np

mesh.point_data['sin'] = np.sin(mesh.point_data['angle'])
mesh.point_data['cos'] = np.cos(mesh.point_data['angle'])

# Subdivision automatically does linear interpolation on the scalar values
subdivided = mesh.subdivide_adaptive(max_n_passes=2, max_tri_area=mesh.area / 1000)  # Whatever area or edge length is appropriate
subdivided.point_data['angle'] = np.arctan2(subdivided.point_data['sin'], subdivided.point_data['cos'])
```

but maybe that will just make things worse

**Reply by davidcortesortuno** [reply]
Interesting idea, but `subdivide_adaptive` does not work on a `RectilinearGrid`

**Reply by banesullivan** [reply]
`subdivide_adaptive` is a filter specific to `PolyData` (deals only with surface data)

**Comment by banesullivan** [comment]
@davidcortesortuno, would you please provide a minimal, reproducible example (mainly looking for a mesh that will highlight the artifcat). I may have some ideas on how to do this, but want an example to work with

**Reply by davidcortesortuno** [reply]
Hi @banesullivan , thanks for looking into this. Yes, here is a minimal example where I colour the dataset according to the in plane azimuthal angle:

```python
import pyvista as pv
import numpy as np

xx = yy = zz = np.linspace(-1, 1, 51)
dataset = pv.RectilinearGrid(xx, yy, zz)

phi = np.arctan2(dataset.points[:, 1], dataset.points[:, 0])
phi[phi < 0] += 2 * np.pi   # set only positive angles
dataset.point_data['angle'] = phi

rad = dataset.points[:, 0] ** 2 + dataset.points[:, 1] ** 2 + dataset.points[:, 2] ** 2
circ = np.zeros_like(dataset.points[:, 0])
circ[rad < 1.] = 1.
dataset.point_data['circle'] = circ

p = pv.Plotter(off_screen=False, notebook=True, window_size=[500, 500])
p.add_mesh(dataset.outline(), color='black')
sphere = dataset.threshold(1., scalars='circle')
p.add_mesh(sphere, opacity=1, style='surface',
           cmap='CET_C1', scalars='angle', show_scalar_bar=False, # use cmocean phase colormap
           # rgba=True, scalars=mpl.cm.hsv(sphere['angle']),
           # interpolate_before_map=False,
           smooth_shading=True, lighting=True, specular=0.6, specular_power=10, ambient=0.)
p.show(jupyter_backend='static')
```

which produces

![image](https://github.com/pyvista/pyvista/assets/12200541/735828a8-2fd0-4ce6-9dbc-1ce9b1997f72)


For this simple case, `interpolate_before_map=False,` gets rid of the interpolation error. However for more complex dataset I still see errors, like the one I showed in my first posts

**Reply by MatthewFlamm** [reply]
One potential solution is to use cell data without interpolation. I used a headless machine to generate so slightly modified your plotting

![image](https://github.com/pyvista/pyvista/assets/39341281/8bdd7dc9-b342-4af3-a525-c06363806c97)


```py
import pyvista as pv
import numpy as np

xx = yy = zz = np.linspace(-1, 1, 51)
dataset = pv.RectilinearGrid(xx, yy, zz)
dataset_centers = dataset.cell_centers()
phi = np.arctan2(dataset_centers.points[:, 1], dataset_centers.points[:, 0])
phi[phi < 0] += 2 * np.pi   # set only positive angles
dataset.cell_data['angle'] = phi

rad = dataset.points[:, 0] ** 2 + dataset.points[:, 1] ** 2 + dataset.points[:, 2] ** 2
circ = np.zeros_like(dataset.points[:, 0])
circ[rad < 1.] = 1.
dataset.point_data['circle'] = circ

p = pv.Plotter(off_screen=True, window_size=[500, 500])
p.add_mesh(dataset.outline(), color='black')
sphere = dataset.threshold(1., scalars='circle')
p.add_mesh(sphere, opacity=1, style='surface',
           cmap='CET_C1', scalars='angle', show_scalar_bar=False, # use cmocean phase colormap
           # rgba=True, scalars=mpl.cm.hsv(sphere['angle']),
           # interpolate_before_map=False,
           smooth_shading=True, lighting=True, specular=0.6, specular_power=10, ambient=0.)
p.show(screenshot="tmp.png")
```

**Reply by davidcortesortuno** [reply]
That's a good quick solution. In my original data I get rid of the errors, however it would be ideal to have interpolation so the cmap looks smooth

![image](https://github.com/pyvista/pyvista/assets/12200541/2ccaeda4-a64d-4e5f-a91d-b46d82c1c2e2)

I guess if we could have some kind of neighbour interpolation between point data, after assigning the cmap colours, it should work

**Reply by banesullivan** [reply, marked as answer]
I think its just a matter of getting the scalar mapping limits (`clim`) to match the period (bounds) of the cycle. So in the case of the example you shared, `clim=[0, 2 * np.pi]`. Additionally, you are correct that `interpolate_before_map=False` is required here too:

```py
           ...
           cmap='CET_C1',
           clim=[0, 2 * np.pi],
           interpolate_before_map=False,
           ...
```

```py
import pyvista as pv
import numpy as np

xx = yy = zz = np.linspace(-1, 1, 51)
dataset = pv.RectilinearGrid(xx, yy, zz)

phi = np.arctan2(dataset.points[:, 1], dataset.points[:, 0])
phi[phi < 0] += 2 * np.pi   # set only positive angles
dataset.point_data['angle'] = phi

rad = dataset.points[:, 0] ** 2 + dataset.points[:, 1] ** 2 + dataset.points[:, 2] ** 2
circ = np.zeros_like(dataset.points[:, 0])
circ[rad < 1.] = 1.
dataset.point_data['circle'] = circ

sphere = dataset.threshold(1., scalars='circle')

p = pv.Plotter(off_screen=False, notebook=True, window_size=[500, 500])
p.add_mesh(dataset.outline(), color='black')
p.add_mesh(sphere, opacity=1, style='surface',
           scalars='angle',
           show_scalar_bar=False,
           cmap='CET_C1',  # use cmocean phase colormap
           clim=[0, 2 * np.pi],
           interpolate_before_map=False,
           smooth_shading=True, lighting=True, specular=0.6, specular_power=10, ambient=0.
          )
p.show(jupyter_backend='static')
```

![download](https://github.com/pyvista/pyvista/assets/22067021/7120b971-9ee0-4bf9-b3d8-516163f36e7e)

So you'll need to figure out what the the period of the other data you're trying this on in order to set this

**Reply by banesullivan** [reply]
For more context on `interpolate_before_map`, see: https://docs.pyvista.org/version/stable/examples/02-plot/interpolate-before-map.html

**Reply by banesullivan** [reply]
I think this would make a great example to add to our docs!

**Reply by banesullivan** [reply]
Here's another, similar scenario: 

```py
import pyvista as pv
import numpy as np

cir = pv.Circle()
cir['cyclical'] = np.arange(cir.n_points)
cyl = cir.extrude([0, 0, 1], capping=False)

cyl.plot(
    cmap='CET_C1',  # use cmocean phase colormap
    clim=[0, cir.n_points],  # must match tht period of the data
    interpolate_before_map=False,  # Critical
    lighting=False,
)
```

![screenshot](https://github.com/pyvista/pyvista/assets/22067021/463d39bb-fe74-4da2-ada0-9c2c79b94a04)

**Reply by banesullivan** [reply]
The reason the `clim` were not automatically set correctly in your example is becuase you are looking at a subset of the full signal: the easiest way to tell was looking at the min/max values:

```py
>>> np.max(sphere['angle'])
6.243206620056296
>>> 2 * np.pi
6.283185307179586
```

This lead to a weird stretching of the colormap that wouldn't cycle properly

This is likely an edge artifact of the `dataset.threshold(1., scalars='circle')` operation
